client/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+# Treode Client Library Interface
+
+The two major components of the Treode client library interface are the [Cache](#Cache) and the [Transaction](#Transaction).  
+
+Caches store key-value pairs retrieved from the server, as well as the read and value timestamps associated with each pair that Treode provides.  Caches are responsible for maintaining a connection to the DB server.
+
+Transactions allows users to locally perform a sequence of database operations and, when finished, optimistically commit the result to the server by writing through the cache.
+
+<a name="Cache"></a>
+# Cache
+
+We provide an overview of the Cache class below, including its primary internal data structure, its public methods, and the form of its HTTP requests and responses.
+
+- Parameters: 
+    1. server: string (required)
+    2. port: string (port=80)
+    3. max\_age: int (max\_age=None)
+    4. no\_cache: bool (no\_cache=False)
+- Public Methods: 
+    1. read(table: string, key: string, max\_age=None, no\_cache=False): value (JSON)
+    2. write(table: string, op\_list: list): status (string), throws StaleException
+- Internal Objects: 
+    1. HTTP library for establishing/managing connection to server (in Python, urllib3)
+    2. map for maintaining cache
+- Exceptions: 
+    1. StaleException
+    2. StaleException(Value_time: int)
```

Comment:
We only one kind of exception. We can determine the value time for read. For write, when the DB gives us a 412, we can use the read_time of that Tx as the value time of the StaleException.  We can do that until we are able to modify the DB to provide a more refined time for the exception.

I think it will be helpful for the user if this Exception also include the read_time of the Tx that produced it.